File: 剑指Offer_24_反转链表/reverseList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # 定义一个缓存列表
        # if head == None:  # 空表节点是None
        #     return head

        # stack = []
        # cur = head
        # while cur:
        #     stack.append(cur)
        #     cur = cur.next

        # for i in range(len(stack)-1):
        #     stack[-i-1].next = stack[-i-2]
        # stack[0].next = None
        # return stack[-1]

        # 迭代
        pro, cur = None, head
        while cur:
            tmp = cur.next
            cur.next = pro
            pro = cur
            cur = tmp
        return pro

File: 剑指Offer_II_24_反转链表/reverseList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    // ====== 迭代 ======
    // let pre = null
    // let cur = head
    // let next
    // while(cur != null){
    //     next = cur.next
    //     cur.next = pre
    //     pre = cur
    //     cur = next
    // }
    // return pre

    // ====== 递归 ======
    // base case
    if(head === null || head.next === null){
        return head
    }
    let newhead = reverseList(head.next)
    head.next.next = head
    head.next = null
    return newhead
};






